### PR TITLE
[DNM] Wounds attempt 2

### PR DESCRIPTION
**Pull request by rye-rice**
## About The Pull Request

Reopen of #305
ESSENTIALLY:
On hit, you have a chance to roll a "wound"
Each wound gives various debuffs which take a certain type of treatment from doctors to fix
This is a placeholder until the baymed port is done

## Why It's Good For The Game
Bone repair surgery is awful
## Changelog
:cl:
add: ports tg wounds
remove: repair machinery surgery
/:cl:

### CODE REVIEW

**Comment by rye-rice** [comment]
ok it works now

File: code/datums/status_effects/debuffs.dm [hunk 1]
```diff
@@ -388,8 +388,19 @@
...
 /datum/status_effect/neck_slice/tick()
 	var/mob/living/carbon/human/H = owner
-	if(H.stat == DEAD || H.bleed_rate <= 8)
+	var/obj/item/bodypart/throat = H.get_bodypart(BODY_ZONE_HEAD)
+	if(H.stat == DEAD || !throat)
 		H.remove_status_effect(/datum/status_effect/neck_slice)
+
+	var/still_bleeding = FALSE
+	for(var/thing in throat.wounds)
+		var/datum/wound/W = thing
```
Comment (Kapu1178) [inline code/datums/status_effects/debuffs.dm]:
```suggestion
	for(var/datum/wound/W as anything in throat.wounds)
```

File: code/game/objects/items/devices/scanners.dm [hunk 2]
```diff
@@ -407,6 +435,70 @@ GENE SCANNER
...
+	if(user.is_blind())
+		to_chat(user, "<span class='warning'>You realize that your scanner has no accessibility support for the blind!</span>")
+		return
+
+	var/render_list = ""
+	for(var/i in patient.get_wounded_bodyparts())
+		var/obj/item/bodypart/wounded_part = i
+		render_list += "<span class='alert ml-1'><b>Warning: Physical trauma[LAZYLEN(wounded_part.wounds) > 1? "s" : ""] detected in [wounded_part.name]</b>"
+		for(var/k in wounded_part.wounds)
+			var/datum/wound/W = k
```
Comment (Kapu1178) [inline code/game/objects/items/devices/scanners.dm]:
```suggestion
		for(var/datum/wound/W as anything in wounded_part.wounds)
```

File: code/game/objects/items/devices/scanners.dm [hunk 3]
```diff
@@ -334,6 +350,18 @@ GENE SCANNER
...
 
+	// Wounds
+	if(iscarbon(M))
+		var/mob/living/carbon/C = M
+		var/list/wounded_parts = C.get_wounded_bodyparts()
+		for(var/i in wounded_parts)
+			var/obj/item/bodypart/wounded_part = i
+			render_list += "<span class='alert ml-1'><b>Warning: Physical trauma[LAZYLEN(wounded_part.wounds) > 1? "s" : ""] detected in [wounded_part.name]</b>"
+			for(var/k in wounded_part.wounds)
+				var/datum/wound/W = k
```
Comment (Kapu1178) [inline code/game/objects/items/devices/scanners.dm]:
```suggestion
			for(var/datum/wound/W as anything in wounded_part.wounds)
```

File: code/game/objects/items/devices/scanners.dm [hunk 4]
```diff
@@ -334,6 +350,18 @@ GENE SCANNER
 		var/tdelta = round(world.time - M.timeofdeath)
 		render_list += "<span class='alert ml-1'><b>Subject died [DisplayTimeText(tdelta)] ago.</b></span>\n"
 
+	// Wounds
+	if(iscarbon(M))
+		var/mob/living/carbon/C = M
+		var/list/wounded_parts = C.get_wounded_bodyparts()
+		for(var/i in wounded_parts)
+			var/obj/item/bodypart/wounded_part = i
```
Comment (Kapu1178) [inline code/game/objects/items/devices/scanners.dm]:
```suggestion
		for(var/obj/item/bodypart/wounded_part as anything in wounded_parts)
```

File: code/datums/status_effects/wound_effects.dm [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+// technically you can have multiple wounds causing limps on the same limb, even if practically only bone wounds cause it in normal gameplay
+	if(left)
+		for(var/thing in left.wounds)
+			var/datum/wound/W = thing
+			slowdown_left += W.limp_slowdown
+			limp_chance_left = max(limp_chance_left, W.limp_chance)
+
+	if(right)
+		for(var/thing in right.wounds)
+			var/datum/wound/W = thing
```
Comment (Kapu1178) [inline code/datums/status_effects/wound_effects.dm]:
```suggestion
		for(var/datum/wound/W as anything in right.wounds)

```

File: code/datums/status_effects/wound_effects.dm [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+
+	slowdown_left = 0
+	slowdown_right = 0
+	limp_chance_left = 0
+	limp_chance_right = 0
+
+// technically you can have multiple wounds causing limps on the same limb, even if practically only bone wounds cause it in normal gameplay
+	if(left)
+		for(var/thing in left.wounds)
+			var/datum/wound/W = thing
```
Comment (Kapu1178) [inline code/datums/status_effects/wound_effects.dm]:
```suggestion
		for(var/datum/wound/W as anything in left.wounds)
```

File: code/datums/elements/embed.dm [hunk 7]
```diff
@@ -173,38 +133,38 @@
...
-	if(!did_embed)
-		payload.failedEmbed()
+	if(istype(payload, /obj/item/shrapnel/bullet))
+		payload.name = P.name
+	payload.embedding = P.embedding
+	payload.updateEmbedding()
+	var/mob/living/carbon/C = hit
+	var/obj/item/bodypart/limb = C.get_bodypart(hit_zone)
+	if(!limb)
+		limb = C.get_bodypart()
```
Comment (Kapu1178) [inline code/datums/elements/embed.dm]:
```suggestion
	limb ||= C.get_bodypart()
```

File: code/__HELPERS/pronouns.dm [hunk 8]
```diff
@@ -262,3 +262,72 @@
 	if((ITEM_SLOT_ICLOTHING in obscured) && skipface)
 		temp_gender = PLURAL
 	return ..()
+
+
+//clothing need special handling due to pairs of items, ie gloves vs a singular glove, shoes, ect.
+/obj/item/clothing/p_they(capitalized, temp_gender)
+	if(!temp_gender)
+		temp_gender = gender
```
Comment (Kapu1178) [inline code/__HELPERS/pronouns.dm]:
```suggestion
	temp_gender ||= gender
```

**Review by Kapu1178 — CHANGES_REQUESTED** [review]

File: code/modules/surgery/bodyparts/bodyparts.dm [hunk 9]
```diff
@@ -82,11 +79,41 @@
...
+	/// If something is currently grasping this bodypart and trying to staunch bleeding (see [/obj/item/grasp_self])
+	var/obj/item/self_grasp/grasped_by
+
+
 /obj/item/bodypart/Initialize()
 	. = ..()
 	name = "[limb_id] [parse_zone(body_zone)]"
 	update_icon_dropped()
 
+	if(bodytype != BODYPART_ORGANIC)
```
Comment (Kapu1178) [inline code/modules/surgery/bodyparts/bodyparts.dm]:
```suggestion
	if(!IS_ORGANIC_LIMB(src))
```

File: code/modules/surgery/bodyparts/bodyparts.dm [hunk 10]
```diff
@@ -163,9 +199,28 @@
 	var/turf/T = get_turf(src)
 	if(IS_ORGANIC_LIMB(src))
 		playsound(T, 'sound/misc/splort.ogg', 50, TRUE, -1)
+	if(current_gauze)
+		qdel(current_gauze)
+	if(current_splint)
+		qdel(current_splint)
+	for(var/obj/item/organ/drop_organ in get_organs())
```
Comment (Kapu1178) [inline code/modules/surgery/bodyparts/bodyparts.dm]:
```suggestion
	for(var/obj/item/organ/drop_organ as anything in get_organs())
```

File: code/modules/surgery/bodyparts/bodyparts.dm [hunk 11]
```diff
@@ -163,9 +199,28 @@
...
 		I.forceMove(T)
 
+///since organs aren't actually stored in the bodypart themselves while attached to a person, we have to query the owner for what we should have
+/obj/item/bodypart/proc/get_organs()
+	if(!owner)
+		return FALSE
+
+	var/list/bodypart_organs
+	for(var/i in owner.internal_organs) //internal organs inside the dismembered limb are dropped.
+		var/obj/item/organ/organ_check = i
```
Comment (Kapu1178) [inline code/modules/surgery/bodyparts/bodyparts.dm]:
```suggestion
	for(var/obj/item/organ/organ_check as anything in owner.internal_organs) //internal organs inside the dismembered limb are dropped.
```

File: code/modules/surgery/bodyparts/bodyparts.dm [hunk 12]
```diff
@@ -196,32 +251,98 @@
...
+	if(owner && wounding_dmg >= WOUND_MINIMUM_DAMAGE && wound_bonus != CANT_WOUND)
+		if(current_gauze)
+			current_gauze.take_damage()
+		if(current_splint)
+			current_splint.take_damage()
+		check_wounding(wounding_type, wounding_dmg, wound_bonus, bare_wound_bonus)
 
-	var/total_damage = brute + burn
+	for(var/i in wounds)
+		var/datum/wound/iter_wound = i
```
Comment (Kapu1178) [inline code/modules/surgery/bodyparts/bodyparts.dm]:
```suggestion
	for(var/datum/wound/iter_wound as anything in wounds)
```

File: code/modules/surgery/bodyparts/bodyparts.dm [hunk 13]
```diff
@@ -233,6 +354,174 @@
...
+		var/datum/wound/loss/dismembering = new
+		dismembering.apply_dismember(src, woundtype, outright=TRUE)
+		return
+
+	// quick re-check to see if bare_wound_bonus applies, for the benefit of log_wound(), see about getting the check from check_woundings_mods() somehow
+	if(ishuman(owner))
+		var/mob/living/carbon/human/human_wearer = owner
+		var/list/clothing = human_wearer.clothingonpart(src)
+		for(var/i in clothing)
+			var/obj/item/clothing/clothes_check = i
```
Comment (Kapu1178) [inline code/modules/surgery/bodyparts/bodyparts.dm]:
```suggestion
		for(var/obj/item/clothing/clothes_check as anything in clothing)
```

File: code/modules/surgery/bodyparts/bodyparts.dm [hunk 14]
```diff
@@ -233,6 +354,174 @@
...
+	for(var/PW in wounds_checking)
+		var/datum/wound/possible_wound = PW
+		var/datum/wound/replaced_wound
+		for(var/i in wounds)
+			var/datum/wound/existing_wound = i
+			if(existing_wound.type in wounds_checking)
+				if(existing_wound.severity >= initial(possible_wound.severity))
+					return
+				else
+					replaced_wound = existing_wound
```
Comment (Kapu1178) [inline code/modules/surgery/bodyparts/bodyparts.dm:454]: This is expensive and badly coded, you're looping through lists a shit load of times.

File: code/modules/surgery/bodyparts/bodyparts.dm [hunk 15]
```diff
@@ -233,6 +354,174 @@
...
+				new_wound = new possible_wound
+				new_wound.apply_wound(src)
+				log_wound(owner, new_wound, damage, wound_bonus, bare_wound_bonus, base_roll)
+			return new_wound
+
+// try forcing a specific wound, but only if there isn't already a wound of that severity or greater for that type on this bodypart
+/obj/item/bodypart/proc/force_wound_upwards(specific_woundtype, smited = FALSE)
+	var/datum/wound/potential_wound = specific_woundtype
+	for(var/i in wounds)
+		var/datum/wound/existing_wound = i
```
Comment (Kapu1178) [inline code/modules/surgery/bodyparts/bodyparts.dm]:
```suggestion
	for(var/datum/wound/existing_wound as anything in wounds)
```

File: code/modules/surgery/bodyparts/bodyparts.dm [hunk 16]
```diff
@@ -233,6 +354,174 @@
...
+  */
+/obj/item/bodypart/proc/check_woundings_mods(wounding_type, damage, wound_bonus, bare_wound_bonus)
+	var/armor_ablation = 0
+	var/injury_mod = 0
+
+	if(owner && ishuman(owner))
+		var/mob/living/carbon/human/H = owner
+		var/list/clothing = H.clothingonpart(src)
+		for(var/c in clothing)
+			var/obj/item/clothing/C = c
```
Comment (Kapu1178) [inline code/modules/surgery/bodyparts/bodyparts.dm]:
```suggestion
		for(var/obj/item/clothing/C as anything in clothing)
```

File: code/modules/surgery/bodyparts/bodyparts.dm [hunk 17]
```diff
@@ -233,6 +354,174 @@
...
+				C.take_damage_zone(body_zone, damage, BURN)
+
+		if(!armor_ablation)
+			injury_mod += bare_wound_bonus
+
+	injury_mod -= armor_ablation
+	injury_mod += wound_bonus
+
+	for(var/thing in wounds)
+		var/datum/wound/W = thing
```
Comment (Kapu1178) [inline code/modules/surgery/bodyparts/bodyparts.dm]:
```suggestion
	for(var/datum/wound/W as anything in wounds)
```

File: code/modules/surgery/bodyparts/bodyparts.dm [hunk 18]
```diff
@@ -649,30 +913,92 @@
...
+  * * replaced- If true, this is being called from the remove_wound() of a wound that's being replaced, so the bandage that already existed is still relevant, but the new wound hasn't been added yet
+  */
+/obj/item/bodypart/proc/update_wounds(replaced = FALSE)
+	var/dam_mul = 1 //initial(wound_damage_multiplier)
 
-/obj/item/bodypart/proc/break_bone()
-	if(!can_break_bone())
+// we can (normally) only have one wound per type, but remember there's multiple types (smites like :B:loodless can generate multiple cuts on a limb)
+	for(var/i in wounds)
+		var/datum/wound/iter_wound = i
```
Comment (Kapu1178) [inline code/modules/surgery/bodyparts/bodyparts.dm]:
```suggestion
	for(var/datum/wound/iter_wound as anything in wounds)
```

File: code/modules/surgery/bodyparts/bodyparts.dm [hunk 19]
```diff
@@ -649,30 +913,92 @@
...
-		receive_damage(rand(1, 3))
-		//1-3 damage every 20 tiles for every broken bodypart.
-		//A single broken bodypart will give you an average of 650 tiles to run before you get a total of 100 damage and fall into crit
+	var/bleed_rate = 0
+	if(generic_bleedstacks > 0)
+		bleed_rate++
+
+	//We want an accurate reading of .len
+	listclearnulls(embedded_objects)
+	for(var/obj/item/embeddies in embedded_objects)
```
Comment (Kapu1178) [inline code/modules/surgery/bodyparts/bodyparts.dm]:
```suggestion
	for(var/obj/item/embeddies as anything in embedded_objects)
```

File: code/modules/surgery/bodyparts/bodyparts.dm [hunk 20]
```diff
@@ -649,30 +913,92 @@
...
+		if(!embeddies.isEmbedHarmless())
+			bleed_rate += 0.5
+
+	for(var/thing in wounds)
+		var/datum/wound/W = thing
```
Comment (Kapu1178) [inline code/modules/surgery/bodyparts/bodyparts.dm]:
```suggestion
	for(var/datum/wound/W as anything in wounds)
```

**Review by Kapu1178 — CHANGES_REQUESTED** [review]
AS ANYTHING

**Comment by rye-rice** [comment]
Remind me to finish this again when all the carbon PRs are done

**Comment by rye-rice** [comment]
i wonder how many merge conflicts this has

**Comment by rye-rice** [comment]
Jesus fucking Christ

**Comment by rye-rice** [comment]
this now has even more conflicts than ever before

**Comment by rye-rice** [comment]
github doesnt even show them all

**Comment by rye-rice** [comment]
confliucts fixed

**Comment by rye-rice** [comment]
when does this never have conflicts bot

**Comment by rye-rice** [comment]
wait this went stale?

**Comment by rye-rice** [comment]
image
ughhhhhhhhhhhhhhhhh

**Comment by Zevotech** [comment]
keep fighting.

**Comment by rye-rice** [comment]
conflicts fixed, again

**Comment by ZephyrTFA** [comment]
As far as I can tell this is not being worked on.
Feel free to make another PR if this is not true

**Pull request closed without merging**